### ISSUE TITLE
🗺️ Phase 1 Roadmap: Noise Cancellation Foundation

### ISSUE THREAD

**Issue by keithah** [keithah/vocana#1]
## 🎯 Project Overview

Vocana is an AI-powered noise cancellation and meeting transcription application for Apple Silicon Mac that combines the functionality of Krisp and Meeter into a single, privacy-focused macOS application. All processing happens locally using Apple's Neural Engine for optimal performance.

### Core Vision
- **Privacy-First**: No data leaves your Mac - all processing is local
- **Apple Silicon Optimized**: Designed specifically for M1+ Macs with Neural Engine acceleration  
- **Real-time Processing**: Sub-10ms latency for noise cancellation
- **System-wide Integration**: Works with any macOS application

---

### Core Components

```
┌─────────────────┐    ┌──────────────────┐    ┌─────────────────┐
│   Menu Bar UI   │    │  Audio Manager   │    │  ML Processing  │
│   (SwiftUI)     │◄──►│  (Core Audio)    │◄──►│  (Core ML)      │
└─────────────────┘    └──────────────────┘    └─────────────────┘
         │                       │                       │
         ▼                       ▼                       ▼
┌─────────────────┐    ┌──────────────────┐    ┌─────────────────┐
│  Settings UI    │    │  Loopback Driver │    │  DeepFilterNet  │
│  (AppKit)       │    │  (Audio Unit)    │    │  (Core ML)      │
└─────────────────┘    └──────────────────┘    └─────────────────┘
         │                       │                       │
         ▼                       ▼                       ▼
┌─────────────────┐    ┌──────────────────┐    ┌─────────────────┐
│  Storage Layer  │    │  System Audio    │    │  Whisper Model  │
│  (SQLite)       │    │  Capture         │    │  (Core ML)      │
└─────────────────┘    └──────────────────┘    └─────────────────┘
```

### Technology Stack
- **Language**: Swift 5.9+
- **Platform**: macOS 14.0+ (Apple Silicon only)
- **Audio**: Core Audio, Audio Unit Extensions, AVAudioEngine
- **ML**: Core ML, Apple Neural Engine, Accelerate framework
- **Storage**: SQLite with GRDB.swift
- **UI**: SwiftUI (menu bar) + AppKit (preferences)
- **Concurrency**: Swift Concurrency (async/await, Actors)

---

### Primary Goals
1. **Establish Core Audio Pipeline**: Implement system-wide audio capture and processing
2. **Deploy Noise Cancellation**: Integrate DeepFilterNet with Core ML optimization
3. **Create Basic UI**: Menu bar interface with recording controls
4. **Validate Performance**: Meet latency and quality benchmarks

#### ✅ In Scope
- Core Audio loopback driver for system audio capture
- DeepFilterNet noise cancellation (Core ML conversion)
- Real-time audio processing pipeline
- Menu bar UI with recording controls
- Basic settings and preferences
- Performance monitoring and logging
- Apple Silicon optimization (Neural Engine)

#### ❌ Out of Scope (Phase 2+)
- Meeting transcription (Whisper integration)
- Calendar integration
- Speaker diarization
- Dashboard interface
- Export functionality
- AI summarization

---

### 1. Core Audio Loopback Driver
**Challenge**: Capturing system audio without requiring user permissions for each application

**Solution**: 
- Implement Audio Unit Extension for system audio capture
- Use `kAudioUnitSubType_HALOutput` for device-level audio access
- Handle sample rate conversion and buffer management
- Implement proper audio session management

**Technical Details**:
```swift
// Audio Unit configuration
let audioUnit = AudioUnitCreateInstance(kAudioUnitType_Output, 
                                       kAudioUnitSubType_HALOutput, 
                                       kAudioUnitManufacturer_Apple)
// Enable input and output
AudioUnitSetProperty(audioUnit, kAudioOutputUnitProperty_EnableIO, 
                    kAudioUnitScope_Input, 1, &flag, 4)
```

### 2. DeepFilterNet to Core ML Conversion
**Challenge**: Converting PyTorch DeepFilterNet model to optimized Core ML format

**Solution**:
- Use `coremltools` Python package for model conversion
- Optimize for Apple Neural Engine where possible
- Implement custom layers for unsupported operations
- Quantize model for INT8 inference on Neural Engine

**Conversion Pipeline**:
```python
import coremltools as ct
import torch

# Load PyTorch model
model = DeepFilterNet.load_from_checkpoint("model.ckpt")
model.eval()

# Trace model
example_input = torch.randn(1, 1, 512)
traced_model = torch.jit.trace(model, example_input)

# Convert to Core ML
mlmodel = ct.convert(
    traced_model,
    inputs=[ct.TensorType(shape=example_input.shape)],
    compute_units=ct.ComputeUnit.ALL
)

# Optimize for Neural Engine
mlmodel = ct.models.neural_network.utils.optimize_for_neural_network(mlmodel)
```

### 3. Real-time Audio Processing Pipeline
**Challenge**: Maintaining sub-10ms latency while processing audio through ML models

**Solution**:
- Implement ring buffer for efficient audio data management
- Use Swift Concurrency with Actors for thread-safe processing
- Optimize buffer sizes (512 samples @ 48kHz = ~10.7ms)
- Implement adaptive processing based on system load

**Architecture**:
```swift
actor AudioProcessor {
    private let ringBuffer = RingBuffer<Float>(capacity: 4096)
    private let deepFilterNet: DeepFilterNetModel
    
    func processAudio(_ samples: [Float]) async -> [Float] {
        // Process in chunks for optimal performance
        let processed = await deepFilterNet.predict(samples)
        return processed
    }
}
```

### 4. Apple Silicon Optimization
**Challenge**: Maximizing performance on M1/M2/M3 chips

**Solution**:
- Leverage Apple Neural Engine for ML inference
- Use Accelerate framework for DSP operations
- Implement SIMD optimizations for audio processing
- Optimize memory allocation and buffer management

**Performance Targets**:
- **Noise Cancellation Latency**: < 10ms
- **CPU Usage**: < 15% on M1 Pro
- **Memory Usage**: < 200MB
- **Neural Engine Utilization**: > 80% when available

---

### Sprint 1: Foundation (Weeks 1-2)
**Goal**: Establish project structure and basic audio capture

**Milestones**:
- [ ] Xcode project setup with proper configuration
- [ ] Core Audio loopback driver implementation
- [ ] Basic audio capture and playback testing
- [ ] Menu bar UI scaffold with SwiftUI
- [ ] Performance monitoring framework

**Deliverables**:
- Working audio capture prototype
- Basic menu bar interface
- Performance benchmarking tools

### Sprint 2: ML Integration (Weeks 3-4)
**Goal**: Integrate DeepFilterNet with Core ML

**Milestones**:
- [ ] DeepFilterNet model conversion to Core ML
- [ ] Core ML integration in Swift
- [ ] Real-time audio processing pipeline
- [ ] Neural Engine optimization
- [ ] Latency optimization and testing

**Deliverables**:
- Working noise cancellation prototype
- Performance benchmarks
- Core ML model files

### Sprint 3: UI and Polish (Weeks 5-6)
**Goal**: Complete user interface and user experience

**Milestones**:
- [ ] Complete menu bar UI implementation
- [ ] Settings and preferences interface
- [ ] Audio device selection and management
- [ ] Real-time performance monitoring UI
- [ ] Error handling and user feedback

**Deliverables**:
- Complete Phase 1 application
- User documentation
- Performance validation report

### Sprint 4: Testing and Release (Weeks 7-8)
**Goal**: Comprehensive testing and release preparation

**Milestones**:
- [ ] Unit and integration testing
- [ ] Performance validation on different Apple Silicon chips
- [ ] User acceptance testing
- [ ] App Store preparation (if applicable)
- [ ] Documentation and deployment guides

**Deliverables**:
- Release-ready application
- Test coverage report
- Performance validation documentation
- Deployment documentation

---

### Technical Metrics
- **Latency**: < 10ms end-to-end audio processing
- **CPU Usage**: < 15% on M1 Pro, < 10% on M2/M3
- **Memory Usage**: < 200MB steady-state
- **Audio Quality**: > 20dB noise reduction
- **Stability**: 99.9% uptime over 24-hour periods

### User Experience Metrics
- **Setup Time**: < 2 minutes from download to working
- **Permission Friction**: Minimal user permissions required
- **Interface Responsiveness**: < 100ms UI response time
- **Error Recovery**: Graceful handling of audio device changes

### Development Quality Metrics
- **Test Coverage**: > 80% code coverage
- **Documentation**: Complete API documentation
- **Performance Monitoring**: Real-time metrics collection
- **Code Quality**: SwiftLint compliance, no warnings

---

### Core Components
- [ ] #1: Core Audio Loopback Driver Implementation
- [ ] #2: DeepFilterNet Core ML Conversion
- [ ] #3: Menu Bar UI Development
- [ ] #4: Real-time Audio Processing Pipeline
- [ ] #5: Apple Silicon Optimization

### Infrastructure
- [ ] #6: Performance Monitoring and Benchmarking
- [ ] #7: Testing Framework Setup
- [ ] #8: Documentation and Deployment Guides
- [ ] #9: CI/CD Pipeline Configuration

### Research and Validation
- [ ] #10: Audio Quality Benchmarking
- [ ] #11: Performance Profiling on Different Apple Silicon Chips
- [ ] #12: User Experience Research and Testing

---

### Target Performance (M1 Pro)
```
Audio Processing Pipeline:
├── Capture Latency: 2ms
├── ML Inference: 5ms  
├── Output Latency: 2ms
└── Total Latency: 9ms

Resource Usage:
├── CPU: 12% (4 performance cores)
├── Neural Engine: 85%
├── Memory: 180MB
└── Power: < 2W additional draw
```

### Quality Metrics
```
Noise Reduction:
├── Background Noise: -20dB to -30dB
├── Voice Preservation: < 1dB distortion
├── Frequency Response: 20Hz - 20kHz (±0.5dB)
└── Dynamic Range: > 90dB
```

---

## 🏁 Next Steps

1. **Immediate Actions** (This Week):
   - Set up development environment and tools
   - Create detailed technical specifications
   - Begin Core Audio driver research and prototyping

2. **Short-term Goals** (Next 2 Weeks):
   - Complete Sprint 1 milestones
   - Establish performance baseline
   - Validate technical approach

3. **Long-term Vision**:
   - Phase 2: Meeting transcription with Whisper
   - Phase 3: Calendar integration and meeting detection
   - Phase 4: AI summarization and advanced features

---

### Technical Risks
- **Core Audio Complexity**: macOS audio system can be challenging
- **Model Conversion**: DeepFilterNet may require custom layer implementations
- **Performance Optimization**: Achieving target latency may require significant optimization

### Mitigation Strategies
- **Incremental Development**: Start with simple audio capture, add complexity gradually
- **Performance Monitoring**: Continuous benchmarking throughout development
- **Fallback Options**: Have alternative approaches ready for each major component

### Dependencies
- **macOS 14.0+**: Required for latest Core Audio features
- **Apple Silicon**: Neural Engine optimization requires M1+ chips
- **Xcode 15.0+**: Latest Swift and Core ML features
- **Python Environment**: For model conversion and testing

---

**Labels**: roadmap, phase-1, documentation, technical, apple-silicon, core-audio, core-ml

**Assignees**: @keithah

**Milestone**: Phase 1 - Noise Cancellation Foundation
